wallet: Decide which coin selection solution to use based on waste metric

achow101 opened the pull request:
Branch and Bound introduced a metric that we call waste. This metric is used as part of bounding the search tree, but it can be generalized to all coin selection solutions, including those with change. As such, this PR introduces the waste metric at a higher level so that we can run both of our coin selection algorithms (BnB and KnapsackSolver) and choose the one which has the least waste. In the event that both find a solution with the same change, we choose the one that spends more inputs.

Also this PR sets the long term feerate to 10 sat/vb rather than using the 1008 block estimate. This allows the long term feerate to be the feerate that we switch between consolidating and optimizing for fees. This also removes a bug where the long term feerate would incorrectly be set to the fallback fee. While this doesn't matter prior to this PR, it does have an effect following this. The long term feerate can be configured by the user through a new `-consolidatefeerate` option.

achow101 committed:
Allow the long term feerate to be configured, default of 10 sat/vb

The long term feerate is really the highest feerate that the user is
comfortable with making consolidatory transactions. This is should thus
be something that can be configured by the user via a new startup option
-consolidatefeerate. The default value is 10 sat/vbyte, chosen
arbitrarily (it seems like a reasonable number).

achow101 committed:
Add waste metric calculation function

achow101 committed:
Use waste metric for deciding which selection to use

Instead of always choosing BnB if it finds a solution, always do both
BnB and KnapsackSolver and choose the one which has the least waste.

achow101 commented:
Note that there will be a followup PR to this which introduces a struct that encapsulates the selection. This struct will also do the waste calculation which will remove the ugly tuples stuff that is currently done to allow for the waste calculation.

achow101 commented:
With #17331 now merged, this is ready for review.

OttoAllmendinger commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,28 @@ CAmount OutputGroup::GetSelectionAmount() const
+    // If we aren't making change, the caller should've set change_cost to 0
+    waste += change_cost;
+
+    // When we are not making change (change_cost  == 0), consider the excess we are throwing away to fees
+    waste += selected_value - target;
```
```
if (change_cost == 0) {
  ..
}
```

achow101 replied:
Fixed.

OttoAllmendinger commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,28 @@ CAmount OutputGroup::GetSelectionAmount() const
+    // When we are not making change (change_cost  == 0), consider the excess we are throwing away to fees
+    waste += selected_value - target;
+
+    return target;
```
`return waste;`

achow101 replied:
Fixed.

sipa commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,49 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+    CAmount bnb_value;
+    bool bnb_ret = SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change, bnb_coins, bnb_value);
+    if (bnb_ret) {
+        results.push_back(std::make_tuple(bnb_coins, bnb_value, /* cost of change */ CAmount(0), nTargetValue));
```
You can std::move `bnb_coins` here to avoid a copy.

achow101 replied:
Done

sipa commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,49 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+    CAmount knapsack_value;
+    bool knapsack_ret = KnapsackSolver(nTargetValue + coin_selection_params.m_change_fee, all_groups, knapsack_coins, knapsack_value);
+    if (knapsack_ret) {
+        results.push_back(std::make_tuple(knapsack_coins, knapsack_value, coin_selection_params.m_cost_of_change, nTargetValue + coin_selection_params.m_change_fee));
```
You can std::move `knapsack_coins` here to avoid a copy.

achow101 replied:
Done

sipa commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,49 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+        CAmount this_waste = GetSelectionWaste(selected_inputs, std::get<2>(result), std::get<3>(result));
+        // Update the best waste either when this waste is better, or when the waste is the same and this result has more inputs
+        if (this_waste < best_waste || (this_waste == best_waste && selected_inputs.size() > setCoinsRet.size())) {
+            setCoinsRet = selected_inputs;
```
It's slightly inefficient to copy the selected inputs into setCoinsRet inside the loop itself, as it may result in multiple useless copies.

As an alternative, you could create a vector of e.g. `(CAmount waste, std::ref<std::set<CInputCoins>> inputs)` pairs by `std::transform`ing the `results` vector, and then returning `*std::min(that vector).second`.

achow101 replied:
Based on this suggestion, I've changed this to do the waste calculation at the time of insertion into `results`. This eliminates the need for a `std::transform`. Then `std::min_element` is used to get the least waste element.

sipa reviewed (commented):
Does this introduce any behavior changes besides the long-term feerate change?

I would think not, as the BNB search always has waste 0, so if it works, it will still always be chosen?

achow101 commented:
> Does this introduce any behavior changes besides the long-term feerate change?
>
> I would think not, as the BNB search always has waste 0, so if it works, it will still always be chosen?

There is a behavior change as BnB can non-zero waste. Part of the waste is the excess thrown away to fees. Because BnB uses a matching window, it is possible for it to have an excess which is included as part of the waste calculation. Additionally, the difference between the fee and the long term fee is usually non-zero, so the waste is also usually non-zero.

The main result of this behavior change is that when target feerate is less than the long term feerate, we will choose the result that consolidates more. This is because the difference between the fee and the long term fee will be negative which generally makes the waste negative. When you have more inputs, the waste will be more negative, and so the most negative solution is used. When the target feerate is greater than the long term feerate, then we will choose the result that pays the least fees. This is because the difference between the fee and the long term fee is positive, so the waste is generally positive. With less inputs, both the waste and fees will be lower, so we end up choosing the solution with lower fees.

This strategy does not necessarily mean that one algorithm will tend to be chosen over another, but rather as a standard way to measure how "good" they are.

sipa commented:
@achow101 Oh, of course. Thanks for reminding me on the point of the waste metric.

jarolrod commented on src/wallet/spend.cpp:
```diff
@@ -672,10 +698,9 @@ bool CWallet::CreateTransactionInternal(
-            coin_selection_params.m_long_term_feerate = GetMinimumFeeRate(*this, cc_temp, nullptr);
+            // Get the long term fee estimate. This estimate is really the fee rate at which we are still willing to consolidate inputs.
+            // By default, we use 10 sat/vbyte.
+            coin_selection_params.m_long_term_feerate = CFeeRate(10, COIN); // Use 10 sat per vbyte
```
What's the rationale behind `10 sat per vbyte`? Why that number?

Seems a bit scary to hard-code a value here.

achow101 replied:
@Xekyo suggested it based on his analysis and experience.

This should really be a configurable option, I just didn't implement that yet.

achow101 replied:
Forgot to mention that this was made a configurable option.

jarolrod commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,43 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+    }
+
+    // Choose the result with the least waste
+    const auto& best_result = std::min_element(results.begin(), results.end(), [](const auto& a, const auto& b){
```
nit
```suggestion
    const auto& best_result = std::min_element(results.begin(), results.end(), [](const auto& a, const auto& b) {
```

achow101 replied:
Done

jarolrod reviewed (commented):
Concept ACK

This is interesting. This allows us to decide which algorithm to use by examining, what can be considered, the opportunity cost of each coin selection algorithm.

Xekyo commented:
> and choose the one which has the least change. In the event that both find a solution with the same change, we choose the one that spends more inputs.

Is it possible that you meant "least waste" and "same waste" instead of change?

achow101 commented:
> > and choose the one which has the least change. In the event that both find a solution with the same change, we choose the one that spends more inputs.
> 
> Is it possible that you meant "least waste" and "same waste" instead of change?

Yes

Xekyo commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -136,6 +136,14 @@ inline std::vector<OutputGroup>& GroupCoins(const std::vector<COutput>& coins)
     return static_groups;
 }
 
+inline std::vector<OutputGroup>& KnapsackGroupOutputs(const CoinEligibilityFilter& filter)
```
Nit: the introduction `KnapsackGroupOutputs` should be mentioned in the commit message. Should be potentially a separate commit.

achow101 replied:
I've mentioned it in the commit message

Xekyo commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -136,6 +136,14 @@ inline std::vector<OutputGroup>& GroupCoins(const std::vector<COutput>& coins)
+{
+    static std::vector<OutputGroup> static_groups;
+    static_groups.clear();
+    static_groups = testWallet.GroupOutputs(vCoins, coin_selection_params, filter, false);
```
How come this uses "GroupOutputs" when BnB uses "GroupCoins"? What's the difference?

achow101 replied:
I've added an explanation in the commit message. The reason is because the `KnapsackSolver` tests involve different eligibility filters which `GroupCoins` does not.

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,30 @@ CAmount OutputGroup::GetSelectionAmount() const
+    // If we aren't making change, the caller should've set change_cost to 0
+    waste += change_cost;
+
+    if (change_cost == 0) {
+        // When we are not making change (change_cost == 0), consider the excess we are throwing away to fees
+        waste += selected_value - target;
+    }
```
This is not just the excess, but `selected_value - target` will also include all input fees paid for the input set. However, we counted the input fees already above in the comparison with the `long_term_fee`.

I think this should be `selected_value - txweight*feerate - sum(recipient outputs)`, or alternatively additionally deduct `sum(coin.m_fee)` for each input for the code as currently written.

A few good tests for the waste metric would be:
- Transaction with change has excess of 0
- Transaction with feerate of 10 sat/vB has waste that is exactly `cost_of_change`+excess (one of which would be zero in the case of having change and the other in the case of not having change)
- Transaction with same count of inputs has higher waster at higher feerate than at lower feerate.

achow101 replied:
Changed to calculated `selected_value` with `coin.effective_value`. This will ensure the fee is subtracted. Also renamed it to `selected_effective_value`.

Xekyo replied:
Ah, very elegant!

Xekyo commented on src/wallet/coinselection.h:
```diff
@@ -166,6 +166,18 @@ struct OutputGroup
 
+/** Compute the waste for this result given the cost of change
+ * waste = change_cost + excess + inputs * (effective_feerate - long_term_feerate)
+ * excess = selected_target - actual_target
```
As pointed out above, this formula would count the input fees again, and here also the change cost.

achow101 replied:
Changed to `selected_effective_value`.

murchandamus reviewed (commented):
Concept ACK, but needs tests for the waste metric calculation.

achow101 committed:
tests: Test GetSelectionWaste

Tests for some waste calculation scenarios

add_coin is modified to allow fee and long term fee to be set for an
added CInputCoin.

murchandamus reviewed (commented):
Good improvements in 6f607eeefb37036d5b0ada89cb3772cbc0de5fb2

Will try to think of more tests and review more carefully before ACK.
Maybe also add a test that out of two transactions with the same excess the one with more inputs has the lower waste at a feerate lower than `long_term_feerate`.

Xekyo commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -658,4 +663,47 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+{
+    CoinSet selection;
+    const CAmount fee = CAmount(100);
+    const CAmount change_cost = CAmount(100);
```
You may want to use different values (preferably not direct multiples) for `fee` and `change_cost` as otherwise some weird interaction might hide a defect, maybe something like 100 and 125.

achow101 replied:
Done

Xekyo commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -658,4 +663,47 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+    BOOST_CHECK_EQUAL(excess, GetSelectionWaste(selection, 0, target));
+    selection.clear();
+
+    // Waste will be greater when fee is greater, but long term fee is the same
```
Perhaps also add a test where the long_term_fee is greater than the fee.

achow101 replied:
Done

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,43 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
+    // Vector of results for use with waste calculation
+    // In order: calculated waste, selected inputs, selected value
+    // TODO: Use a struct representing the selection result
+    std::vector<std::tuple<CAmount, std::set<CInputCoin>, CAmount>> results;
```
You may want to be a bit more explicit here whether the third value is the selected value, the selected effective value, or the target value. (I assume it's the first, but we've been juggling all of these, being more precise wouldn't hurt.)

achow101 replied:
Done

meshcollider replied:
Do you anticipate having more coin selection algorithms in future? Why bother with the vector etc. rather than just direct comparison between the two results?

achow101 replied:
Yes, #17526 will be updated to add another result to this vector.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,43 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
+        return false;
+    }
+
+    // Choose the result with the least waste
```
Clearly, this function does not only select by waste, but has a fallback tiebreaker. ;)

achow101 replied:
Done

murchandamus reviewed (commented):
Some minor points and a suggestion for another test.

achow101 committed:
tests: Use SelectCoinsBnB directly instead of AttemptSelection

Instead of calling AttemptSelection with the hopes/instruction that it
uses BnB, call SelectCoinsBnB directly to test it.

achow101 committed:
tests: Add KnapsackGroupOutputs helper function

In order to change the KnapsackSolver tests to call KnapsackSolver, we
need KnapsackGroupOutputs to create the OutputGroups filtered with the
filter criteria.

achow101 committed:
scripted-diff: tests: Use KnapsackSolver directly

When doing the coin selector tests for KnapsackSolver, call it directly
instead of using AttemptSelection and hoping/instructing it uses
KnapsackSolver.

-BEGIN VERIFY SCRIPT-
sed -i 's/testWallet\.AttemptSelection( /KnapsackSolver(/g' src/wallet/test/coinselector_tests.cpp
sed -i 's/testWallet\.AttemptSelection(/KnapsackSolver(/g' src/wallet/test/coinselector_tests.cpp
sed -i 's/, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params)/, KnapsackGroupOutputs(filter_standard), setCoinsRet, nValueRet)/g' src/wallet/test/coinselector_tests.cpp
sed -i 's/, filter_confirmed, vCoins, setCoinsRet, nValueRet, coin_selection_params)/, KnapsackGroupOutputs(filter_confirmed), setCoinsRet, nValueRet)/g' src/wallet/test/coinselector_tests.cpp
sed -i 's/, filter_standard_extra, vCoins, setCoinsRet, nValueRet, coin_selection_params)/, KnapsackGroupOutputs(filter_standard_extra), setCoinsRet, nValueRet)/g' src/wallet/test/coinselector_tests.cpp
sed -i 's/BOOST_CHECK( /BOOST_CHECK(/g' src/wallet/test/coinselector_tests.cpp
-END VERIFY SCRIPT-

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -280,14 +288,14 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
     add_coin(1);
     vCoins.at(0).nInputBytes = 40; // Make sure that it has a negative effective value. The next check should assert if this somehow got through. Otherwise it will fail
-    BOOST_CHECK(!testWallet.AttemptSelection( 1 * CENT, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params_bnb));
+    BOOST_CHECK(!SelectCoinsBnB(GroupCoins(vCoins), 1 * CENT, coin_selection_params_bnb.m_cost_of_change, setCoinsRet, nValueRet));
```
Done

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,30 @@ CAmount OutputGroup::GetSelectionAmount() const
+
+    if (change_cost == 0) {
+        // When we are not making change (change_cost == 0), consider the excess we are throwing away to fees
+        waste += selected_effective_value - target;
```
Done

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -166,6 +166,18 @@ struct OutputGroup
 
+/** Compute the waste for this result given the cost of change
+ * waste = change_cost + excess + inputs * (effective_feerate - long_term_feerate)
+ * excess = selected_effective_value - actual_target
```
Changed to just `target`.

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -166,6 +166,18 @@ struct OutputGroup
+ * @param[in] inputs The selected inputs
+ * @param[in] change_cost The cost of creating change and spending it in the future.
+ * @param[in] target The amount targeted by the coin selection algorithm.
+ * @return The waste
```
Done

Xekyo commented on src/wallet/coinselection.h:
```diff
@@ -166,6 +166,19 @@ struct OutputGroup
+ * @param[in] real_value Whether to use the input's real value (when true) or the effective value (when false)
+ * @return The waste
+ */
+CAmount GetSelectionWaste(const std::set<CInputCoin>& inputs, const CAmount change_cost, const CAmount target, bool real_value = false);
```
I'd have a slight preference for a more specific "use_effective_value" than "real_value" with an inverted bool value.
```suggestion
CAmount GetSelectionWaste(const std::set<CInputCoin>& inputs, const CAmount change_cost, const CAmount target, bool use_effective_value = true);
```

Talkless replied:
Maybe worth adding `[[nodiscard]]` for a new `GetSelectionWaste`?

achow101 replied:
Done both suggestions.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,44 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
+    // Choose the result with the least waste
+    // If the waste is the same, choose the one which spends more inputs.
+    const auto& best_result = std::min_element(results.begin(), results.end(), [](const auto& a, const auto& b){
+        return std::get<0>(a) < std::get<0>(b) || (std::get<0>(a) == std::get<0>(b) && std::get<1>(a).size() > std::get<1>(b).size());
```
Note to reviewers: this looks much better with #22019

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -658,10 +685,9 @@ bool CWallet::CreateTransactionInternal(
-    coin_selection_params.m_long_term_feerate = GetMinimumFeeRate(*this, cc_temp, nullptr);
+    // Get the long term fee estimate. This estimate is really the fee rate at which we are still willing to consolidate inputs.
+    // By default, we use 10 sat/vbyte.
+    coin_selection_params.m_long_term_feerate = CFeeRate(10, COIN); // Use 10 sat per vbyte
```
Warrants updating comment in `coinselection.h`

nit: duplicate comment with the one just one line above

achow101 replied:
No longer relevant with `-consolidatefeerate` option.

S3RK commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -650,4 +663,67 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+    BOOST_CHECK_EQUAL(fee_diff * -2 + excess, waste_nochange2);
+    BOOST_CHECK_LT(waste_nochange2, waste_nochange1);
+    selection.clear();
+
```
nit: Maybe add a test for zero waste case?

achow101 replied:
Done

S3RK reviewed (approved):
Code Review ACK  8a0a1b3

I'm woking on a functional test to verify the overall flow. The idea is to craft a specific coin set, set the fee manually to A) below B) above long_term_fee and verify that we've got more inputs in case A

ryanofsky commented:
> I'm woking on a functional test to verify the overall flow

This reminds me to beg for review of #22155. Functional tests for coin selection are great but for some parts unit tests are more convenient, so I really want to add this wallet spend test module and get some tests in there.

S3RK commented on src/dummywallet.cpp:
```diff
@@ -28,6 +28,7 @@ void DummyWalletInit::AddWalletOptions(ArgsManager& argsman) const
         "-addresstype",
         "-avoidpartialspends",
         "-changetype",
+        "-discardfee=<amt>",
```
AssertionError: Please add {'-consolidatefeerate'} to the hidden args in DummyWalletInit::AddWalletOptions

achow101 replied:
Oops, fixed.

S3RK commented on src/wallet/init.cpp:
```diff
@@ -45,6 +45,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-addresstype", strprintf("What type of addresses to use (\"legacy\", \"p2sh-segwit\", or \"bech32\", default: \"%s\")", FormatOutputType(DEFAULT_ADDRESS_TYPE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-avoidpartialspends", strprintf("Group outputs by address, selecting many (possibly all) or none, instead of selecting on a per-output basis. Privacy is improved as addresses are mostly swept with fewer transactions and outputs are aggregated in clean change addresses. It may result in higher fees due to less optimal coin selection caused by this added limitation and possibly  ...
     argsman.AddArg("-changetype", "What type of change to use (\"legacy\", \"p2sh-segwit\", or \"bech32\"). Default is same as -addresstype, except when -addresstype=p2sh-segwit a native segwit output is used when sending to a native segwit address)", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
+    argsman.AddArg("-consolidatefeerate=<amt>", strprintf("The fee rate (in %s/kvB) that indicates the highest feerate you are comfortable making consolidation transactions at (default: %s). ", CURRENCY_UNIT, FormatMoney(DEFAULT_CONSOLIDATE_FEERATE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
nit: unnecessary space after the dot in the description

achow101 replied:
Fixed.

S3RK reviewed (commented):
Re-ACK 1ce0c97

Compared to 8a0a1b3 the first commit is replaced to make consolidating fee rate configurable.

fjahr reviewed (commented):
Concept ACK

Looks good already at the first look but I will need to do another pass with a focus on the tests.

GeneFerneau commented:
Approach ACK [1ce0c97](https://github.com/bitcoin/bitcoin/pull/22009/commits/1ce0c97c7a531070328d4d742a32b31e1e0bdce5)

Did a manual review, everything looks good. Tangentially, doing a deeper review of `KnapsackSolver`, but use looks consistent with other uses across the codebase.

Still running tests locally, and trying to think of more edge cases. The existing tests look pretty throrough.

Talkless commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,44 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
-        return true;
+    std::set<CInputCoin> bnb_coins;
+    CAmount bnb_value;
+    bool bnb_ret = SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change, bnb_coins, bnb_value);
```
Is it necessary to bring variable in to scope that is used only for `if`? Probably personal preference, but adding variable automatically "makes you" (a reviewer) to look for it's uses elsewhere, including possible re-assignemnt as it it's nor marked as `const`.

Please consider adding `const` or just remove that variable and "`if`" the `SelectCoinsBnB` result directly.

achow101 replied:
This variable was used in a previous revision. Changed to use the in the `if` directly.

Talkless commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,44 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
-    return KnapsackSolver(nTargetValue + coin_selection_params.m_change_fee, all_groups, setCoinsRet, nValueRet);
+    std::set<CInputCoin> knapsack_coins;
+    CAmount knapsack_value;
+    bool knapsack_ret = KnapsackSolver(nTargetValue + coin_selection_params.m_change_fee, all_groups, knapsack_coins, knapsack_value);
```
Same as per `bool bnb_ret` comment.

achow101 replied:
Done

Talkless commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -136,6 +141,14 @@ inline std::vector<OutputGroup>& GroupCoins(const std::vector<COutput>& coins)
+inline std::vector<OutputGroup>& KnapsackGroupOutputs(const CoinEligibilityFilter& filter)
+{
+    static std::vector<OutputGroup> static_groups;
+    static_groups.clear();
```
Why `.clear()`? It does not change the capacity (actual internal buffer size, if that was in mind). 

Just next line completely re-assigns `static_groups` and it seems that should be enough?

achow101 replied:
This was copied from one of the other functions, but indeed is unnecessary, so I have removed it.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -48,14 +48,19 @@ static void add_coin(const CAmount& nValue, int nInput, std::vector<CInputCoin>&
+    set.insert(coin);
 }
 
+
```
Removed

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -658,4 +663,67 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+    BOOST_CHECK_LT(waste3, waste1);
+    selection.clear();
+
+    // Waste without change is the excesss and difference between fee and long term fee
```
Fixed.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -658,4 +663,67 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+    BOOST_CHECK_EQUAL(fee_diff * 2 + change_cost, waste1);
+    selection.clear();
+
+    // Waste without change is the excesss and difference between fee and long term fee
```
Fixed.

achow101 commented:
Also rebased to pick up #22686

achow101 commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -543,7 +543,7 @@ def test_locked_wallet(self):
         self.nodes[1].getrawchangeaddress()
         inputs = []
-        outputs = {self.nodes[0].getnewaddress():1.09999500}
+        outputs = {self.nodes[0].getnewaddress():1.19999500}
```
Note: I believe this change is related to #22686. What I think was happening in previous revisions of this PR is that due to that bug, `KnapsackSolver` would find a solution without change that doesn't actually work, and because of that, it would also have worse waste than the BnB solution. We don't see this in master because there is a BnB solution and in master, we default to the BnB solution. This PR, combined with #22686 allows `KnapsackSolver` to find a solution with change that happens to have less waste, so it was being chosen, causing the test to fail at the `fundrawtransaction` below. So this value is being changed to just under 1.2 which is where an optimal input set is being found.

S3RK replied:
I recompiled the old version of this PR and confirmed that

murchandamus reviewed (commented):
Re: 810b7fe7ba9504c36de3e190a8c901e2220dd77f
LGTM, just got some more nits for you, I guess.

achow101 commented on src/wallet/init.cpp:
```diff
@@ -45,6 +45,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-addresstype", strprintf("What type of addresses to use (\"legacy\", \"p2sh-segwit\", or \"bech32\", default: \"%s\")", FormatOutputType(DEFAULT_ADDRESS_TYPE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-avoidpartialspends", strprintf("Group outputs by address, selecting many (possibly all) or none, instead of selecting on a per-output basis. Privacy is improved as addresses are mostly swept with fewer transactions and outputs are aggregated in clean change addresses. It may result in higher fees due to less optimal coin selection caused by this added limitation and possibly  ...
     argsman.AddArg("-changetype", "What type of change to use (\"legacy\", \"p2sh-segwit\", or \"bech32\"). Default is same as -addresstype, except when -addresstype=p2sh-segwit a native segwit output is used when sending to a native segwit address)", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
+    argsman.AddArg("-consolidatefeerate=<amt>", strprintf("The fee rate (in %s/kvB) that indicates the highest feerate you are comfortable making consolidation transactions at (default: %s).", CURRENCY_UNIT, FormatMoney(DEFAULT_CONSOLIDATE_FEERATE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
Changed.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -659,10 +659,8 @@ bool CWallet::CreateTransactionInternal(
-    CCoinControl cc_temp;
-    cc_temp.m_confirm_target = chain().estimateMaxBlocks();
-    coin_selection_params.m_long_term_feerate = GetMinimumFeeRate(*this, cc_temp, nullptr);
+    // Set the long term fee estimate to the wallet's consolidate feerate
```
Changed.

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,31 @@ CAmount OutputGroup::GetSelectionAmount() const
+
+CAmount GetSelectionWaste(const std::set<CInputCoin>& inputs, const CAmount change_cost, const CAmount target, bool use_effective_value)
+{
+    // If the result is empty, then selection failed. Don't use this one, so set waste to very high
```
Originally the intent was to always run this function regardless of success or failure. But we won't be doing that, so it's fine to assert here. Done.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -659,10 +659,8 @@ bool CWallet::CreateTransactionInternal(
-    cc_temp.m_confirm_target = chain().estimateMaxBlocks();
-    coin_selection_params.m_long_term_feerate = GetMinimumFeeRate(*this, cc_temp, nullptr);
+    // Set the long term fee estimate to the wallet's consolidate feerate
+    coin_selection_params.m_long_term_feerate = m_consolidate_feerate;
```
Changed.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -49,12 +49,16 @@ static void add_coin(const CAmount& nValue, int nInput, std::vector<CInputCoin>&
     tx.vout[nInput].nValue = nValue;
-    set.emplace(MakeTransactionRef(tx), nInput);
+    CInputCoin coin(MakeTransactionRef(tx), nInput);
+    coin.effective_value -= fee;
```
Changed.

Xekyo commented:
ACK 15fbe72ede62b9a3686fc000381c1d0fa1807815

instagibbs commented:
>and choose the one which has the least change. In the event that both find a solution with the same change, we choose the one that spends more inputs.

> Is it possible that you meant "least waste" and "same waste" instead of change?

>Yes

Could you please update the OP with accurate language? I was struggling to understand until I read everything here.

instagibbs commented on src/wallet/wallet.h:
```diff
@@ -71,6 +71,8 @@ static const CAmount DEFAULT_FALLBACK_FEE = 0;
 //! -mintxfee default
 static const CAmount DEFAULT_TRANSACTION_MINFEE = 1000;
+//! -consolidatefee default
+static const CAmount DEFAULT_CONSOLIDATE_FEERATE = 10000; // 10 sat/vbyte
```
definitely not going to nitpick this but why 10? :)

achow101 replied:
It seemed like a reasonable number

Xekyo replied:
Given that we have gotten that question twice again in the last 24h, it would probably be best to add context for that number in the commit messages or code somewhere.

achow101 replied:
I added to the commit message that it was chosen arbitrarily.

instagibbs commented:
concept ACK

achow101 commented:
Updated OP

meshcollider commented on src/wallet/wallet.h:
```diff
@@ -71,6 +71,8 @@ static const CAmount DEFAULT_FALLBACK_FEE = 0;
 static const CAmount DEFAULT_DISCARD_FEE = 10000;
 //! -mintxfee default
 static const CAmount DEFAULT_TRANSACTION_MINFEE = 1000;
+//! -consolidatefee default
```
nit: `-consolidatefeerate`

meshcollider reviewed (commented):
utACK 1858364a949ea8345df5e845e7d1433b7cd5ffd0

Xekyo commented:
@meshcollider: I think there are currently three (Knapsack, BnB, and SRD), and it might be interesting to add a couple more. E.g. smallest-first could be an interesting strategy to create transactions at `minRelayTxFee`, or at very high feerates, largest-first could perhaps be preferred if it's smaller.

Xekyo commented:
As a follow-up to my comment about the constructor of `CInputCoin` initializing `effective_value` with `nValue` and the subsequent update of the `effective_value` to the correct magnitude: I ran the functional tests after commenting out the line that initializes `effective_value` in the `CInputCoin` with Valgrind and all tests passed, so I think that it might be possible to make `CInputCoin` const and thus force setting the correct value from the get-go. What do you think about that?

achow101 commented:
@Xekyo I'm not quite sure what you mean. The members of `CInputCoin` can't be const because the constructor has some logic before they can be set.

S3RK commented:
utACK per `git range-diff fdd80b0a5 1ce0c97 1858364`. The only changes are related to comments in the PR

Xekyo commented:
As discussed, let's leave it for a follow-up PR.

Pankylove99 commented:
Good evening from Italy I am a member here please how can I  come into bitcoin business...i need your help bro God bless you 

Sent from Yahoo Mail on Android 
 
  On Mon, 23 Aug 2021 at 16:37, Mark ***@***.***> wrote:   


As discussed, let's leave it for a follow-up PR.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

glozow commented on src/wallet/spend.cpp:
```diff
@@ -357,17 +357,42 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
+    std::set<CInputCoin> knapsack_coins;
+    CAmount knapsack_value;
+    if (KnapsackSolver(nTargetValue + coin_selection_params.m_change_fee, all_groups, knapsack_coins, knapsack_value)) {
+        results.push_back(std::make_tuple(GetSelectionWaste(knapsack_coins, coin_selection_params.m_cost_of_change, nTargetValue + coin_selection_params.m_change_fee, !coin_selection_params.m_subtract_fee_outputs), std::move(knapsack_coins), knapsack_value));
```
in 1858364a949ea8345df5e845e7d1433b7cd5ffd0 _Use waste metric for deciding which selection to use_:
`emplace_back` to construct in place
break up the line for readability
(same with the BnB solving portion above)
```suggestion
        const auto waste = GetSelectionWaste(knapsack_coins, coin_selection_params.m_cost_of_change, nTargetValue + coin_selection_params.m_change_fee, !coin_selection_params.m_subtract_fee_outputs);
        results.emplace_back(std::make_tuple(waste, std::move(knapsack_coins), knapsack_value));
```

achow101 replied:
Done

glozow reviewed (commented):
ACK 1858364a949ea8345df5e845e7d1433b7cd5ffd0, some minor suggestions

Waste metric changes make sense and look correct to me. I'm not able to determine whether 10 sat/vbyte is a good consolidation feerate, though as an individual user, I would accept consolidating UTXOs at 10sat/vbyte or lower. While testing, I added some lines to compare the long term fee estimate with 10, and they at least don't seem drastically different.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -137,6 +141,13 @@ inline std::vector<OutputGroup>& GroupCoins(const std::vector<COutput>& coins)
+inline std::vector<OutputGroup>& KnapsackGroupOutputs(const CoinEligibilityFilter& filter)
+{
+    static std::vector<OutputGroup> static_groups;
+    static_groups = testWallet.GroupOutputs(vCoins, coin_selection_params, filter, false);
```
Done

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -166,6 +166,19 @@ struct OutputGroup
+ * @param[in] change_cost The cost of creating change and spending it in the future. Only used if there is change. Must be 0 if there is no change.
+ * @param[in] target The amount targeted by the coin selection algorithm.
+ * @param[in] use_effective_value Whether to use the input's effective value (when true) or the real value (when false).
+ * @return The waste
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -71,6 +71,8 @@ static const CAmount DEFAULT_FALLBACK_FEE = 0;
 //! -mintxfee default
 static const CAmount DEFAULT_TRANSACTION_MINFEE = 1000;
+//! -consolidatefee default
+static const CAmount DEFAULT_CONSOLIDATE_FEERATE = 10000; // 10 sat/vbyte
```
Changed the initialization I don't think static asserts are necessary.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -659,10 +684,8 @@ bool CWallet::CreateTransactionInternal(
-    cc_temp.m_confirm_target = chain().estimateMaxBlocks();
-    coin_selection_params.m_long_term_feerate = GetMinimumFeeRate(*this, cc_temp, nullptr);
+    // Set the long term feerate estimate to the wallet's consolidate feerate
+    coin_selection_params.m_long_term_feerate = m_consolidate_feerate;
```
Done

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,29 @@ CAmount OutputGroup::GetSelectionAmount() const
     return m_subtract_fee_outputs ? m_value : effective_value;
 }
+
+CAmount GetSelectionWaste(const std::set<CInputCoin>& inputs, const CAmount change_cost, const CAmount target, bool use_effective_value)
```
Done

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -341,3 +341,29 @@ CAmount OutputGroup::GetSelectionAmount() const
+        // When we are not making change (change_cost == 0), consider the excess we are throwing away to fees
+        assert(selected_effective_value >= target);
+        waste += selected_effective_value - target;
+    }
```
Done

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -651,4 +662,73 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+    const CAmount fee_diff = CAmount(40);
+    const CAmount in_amt = 3 * COIN;
+    const CAmount target = 2 * COIN;
+    const CAmount excess = in_amt - fee * 2 - target;
```
Done

achow101 commented:
Had to rebase due to a hidden merge conflict.

prayank23 commented:
> Also this PR sets the long term feerate to 10 sat/vb rather than using the 1008 block estimate.

Initially this didn't look like the right approach but then found this option is configurable.

According to this comment it was chosen arbitrarily: https://github.com/bitcoin/bitcoin/pull/22009#discussion_r692371546

According to this comment Murch suggested it based on some analysis : https://github.com/bitcoin/bitcoin/pull/22009#discussion_r647594412  

1. Can you share the analysis used for this value?
2. How does this affect Coin Selection?

glozow reviewed (commented):
re ACK  fe47558 via `git range-diff 1858364...fe47558`, thanks for accepting suggestions

murchandamus reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/commit/fe47558acd365264778af5eefd61f0955e168c6f via `git range-diff 15fbe72...fe47558`

Xekyo commented:
reACK 86beee0 via git range-diff fe47558...86beee0

meshcollider commented:
re-utACK 86beee05795216738f51fa744539336503c26fd9

glozow reviewed (commented):
post merge ACK 86beee0 via `git range-diff fe47558...86beee0`
changes were addressing https://github.com/bitcoin/bitcoin/pull/22009#discussion_r697503379, https://github.com/bitcoin/bitcoin/pull/22009#discussion_r697426583, https://github.com/bitcoin/bitcoin/pull/22009#discussion_r692588236

rajarshimaitra commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -651,4 +662,73 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
+    BOOST_CHECK_LT(waste_nochange2, waste_nochange1);
+    selection.clear();
+
+    // 0 Waste only when fee == long term fee, no change, and no excess
+    add_coin(1 * COIN, 1, selection, fee, fee);
+    add_coin(2 * COIN, 2, selection, fee, fee);
+    const CAmount exact_target = in_amt - 2 * fee;
+    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, 0, exact_target));
```
As it was discussed in the [review club](https://github.com/bitcoin/bitcoin/pull/22938), there are more scenarios when the waste can be zero. Thus the "only" doesn't seem appropriate here. 

I went ahead and opened another PR fixing this with addition of those extra cases. https://github.com/bitcoin/bitcoin/pull/22938. 

Hope that is helpful.

rajarshimaitra reviewed (commented):
Post merge tACK
